### ISSUE TITLE
Implementation of layerdrop

### ISSUE THREAD

**Issue by prajjwal1** [pytorch/fairseq#1667]
I wanted to learn more about the implementation of layerdrop. As per the paper, there are three types of layerdrop. I think fairseq uses `every other`. The rest two maybe released later as mentioned.  I have seen the layerdrop being used [here](https://github.com/pytorch/fairseq/blob/f8180af3555ba8b185af95f71b14863ea9146cc3/fairseq/modules/transformer_sentence_encoder.py#L220).  I wanted to know that we're not specifying any layers here. So is layerdrop being performed in the same manner as dropout like specifying the drop pct and it randomly drops some layers. As per the code,
`x, _ = layer(x, self_attn_padding_mask=padding_mask)
`
it seems that the inputs are passed through all the layers. Could you also please explain the implementation of layerdrop ?

**Comment by myleott** [comment]
Training is always implemented as random skipping of layers, usually with p=0.2 according to Section 4 of the paper: "We set the LayerDrop rate p to 0.2".

During inference one can drop with different strategies, such as every other, first/last K, etc. This is implemented by rewriting the checkpoint on load: https://github.com/pytorch/fairseq/blob/9ebcd6554daab7bac4948de49aeb85bfff81d876/fairseq/checkpoint_utils.py#L365-L446.

**Comment by prajjwal1** [comment]
Could you also provide the implementation for data driven layerdrop ?
